docs/hibernating-clusters.md

```diff
@@ -0,0 +1,121 @@
+
+Hive can automate the process of stopping/starting clusters via its API by allowing the user to
+set a the desired state of the cluster in the ClusterDeployment spec. Both API and controller
+changes are required to support this feature.
+
+## API Changes
+
+The ClusterDeploymentSpec should allow setting whether machines are in a running state or in
+a hibernating state.
+
+```go
+type ClusterDeploymentSpec struct {
+  // ... other fields
+
+  // Hibernating indicates whether a cluster should be hibernating. Default is false.
+  Hibernating  `json:"hibernating,omitempty"`
+}
+```
+
+On the status side, a cluster's state is reflected by the following condition types:
+
+```go
+const (
+  // ... other constants
+
+  // ClusterStartingConditionType is set when the ClusterDeployment is transitioning
+  // from a hibernating state to a running state.
+  ClusterStartingConditionType ClusterDeploymentConditionType = "ClusterStarting"
+
+  // ClusterRunningConditionType is set when the ClusterDeployment is in a running and
+  // ready state.
+  ClusterRunningConditionType ClusterDeploymentConditionType = "ClusterRunning"
+
+  // ClusterStoppingConditionType is set when the ClusterDeployment is transitioning
+  // from a running state to a hibernating state.
+  ClusterStoppingConditionType ClusterDeploymentConditionType = "ClusterStopping"
+
+  // ClusterHibernatingConditionType is set when the ClusterDeployment has transitioned
+  // to a hibernating state.
+  ClusterHibernatingConditionType ClusterDeploymentConditionType = "ClusterHibernating"
+
+  // HibernationNotSupportedConditionType is set when a ClusterDeployment's hibernating
+  // property is set to true, but the cluster has an OpenShift version that doesn't support hibernating
+  // (< 4.4.8).
+  HibernationNotSupportedConditionType ClusterDeploymentConditionType = "HibernationNotSupported"
+)
+```
+
+## Controller Changes
+
+### Cluster Hibernation Controller
+The cluster hibernation controller is a new controller that will watch ClusterDeployments and ensure that the target cluster's
+machines reflect the hibernating state specified in the ClusterDeployment spec.
+
+High level flow of controller logic:
+
+![hibernation_controller](hibernation_controller.png)
+
+The controller uses the actuator pattern to power up and power down machine instances as well as determining
+machine state (running or stopped).
+
+#### Selecting Cluster Machines
+
+Option 1 (Preferred):
+The hibernation controller relies on the actuator to select machines used by the cluster.
+The actuator, given a ClusterDeployment's InfraID selects machines using a method appropriate to the cloud provider (tags/name prefix/resource group).
+
+Option 2:
+The hibernation controller uses the machine API on the target cluster to determine which machines belong to the cluster. It then stores the machine IDs
+in the clusterdeployment (or a separate CR), then uses those machine IDs to start the cluster again.
```

Comment:
Would a configmap be preferable here? We just want to store some data, I don't know as we necessary need API level commitment here.

Reply:
Ack a configmap would work. However, would you prefer this option to using tags/name ?

Reply:
Option 1 seems more resilient to me.